src/node_id.rs

```diff
@@ -0,0 +1,334 @@
+            return Err(NodeIdGenError::PubkeyHashZero);
+        } else {
+            return Ok(new_node);
+        }
+    }
+
+    /// Generate NodeId with Resistance
+    ///
+    /// Requires disco::hash(public_key) to be have `difficulty` number of trailing zeros
+    /// TODO: hash should be of some changing shared state for better sybil resistance
+    pub fn hard_generate(difficulty: usize, timeout: usize) -> Result<NodeId, NodeIdGenError> {
+        let clock = SteadyTime::now();
+        loop {
+            let new_id = NodeId::generate()?;
+            // default trailing zeros (remove `rev()` for leading zeros)
+            //let disco_iter = new_id.discohash.iter().rev();
+            let mut success = true;
+            for i in 0..difficulty {
+                if new_id.discohash.get(i).unwrap() != &0u8 {
+                    success = false;
+                }
+            }
+            if success {
+                return Ok(new_id);
+            }
+            //                           converts usize into i64 and panics if doesn't fit
+            // TODO: consider if this is best or if I should just pass i64 as an argument
+            if SteadyTime::now() - clock > Duration::seconds(timeout.try_into().unwrap()) {
+                return Err(NodeIdGenError::HardGenTimeOut);
+            }
+        }
+    }
+
+    #[inline]
+    pub fn to_base58(&self) -> String {
+        bs58::encode(self.discohash.as_slice()).into_string()
+    }
+
+    #[inline]
+    fn digest(&self) -> &[u8] {
+        &self.discohash.as_slice()
+    }
+
+    // TODO: better verification that `data` is a valid `NodeId`
+    // would require logic from `is_public_key`
+    #[inline]
+    fn from_bytes(data: Vec<u8>) -> Result<NodeId, ParseError> {
+        if data.len() != 32 {
+            return Err(ParseError);
+        }
+        let new_node = NodeId { discohash: data };
+        Ok(new_node)
+    }
+
+    #[inline]
+    fn is_public_key(&self, pubkey: PublicKey) -> bool {
+        let ret_val = true;
+        let mut counter = 0;
+        let pk_hash = hash(pubkey.as_bytes(), 32);
+        &pk_hash == &self.discohash
+    }
+
+    /// For Testing Purposes Only
+    ///
+    /// Note: does not generate keypair, just generates a random byte array
+    #[inline]
+    pub fn random(output_len: usize) -> NodeId {
+        NodeId {
+            discohash: DiscoHash::random(output_len),
+        }
+    }
+
+    /// Useful for distance metric and node_id generation checks
+    pub fn is_zero(&self) -> bool {
+        self.discohash.iter().all(|d| *d == 0)
+    }
+
+    /// Verify Signature
+    ///
+    /// verify message signature made with the public key associated with this NodeId
+    pub fn verify_msg(message: &[u8], signature: &[u8], public_key: &[u8]) -> bool {
+        // change receive type to an error on verifying signatures
+
+        // would need to use the sign method in Node
+        // or take a signature
+        todo!();
```

Comment:
todo

Reply:
should this be done before this is merged? Shouldn't be too hard, but I'd need to implement `sign_message` on `config::NodeHandler`

Reply:
was wrong about this, just implemented `verify` and added a test to sign and verify...